File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque
dx=[-1,1,0,0]
dy=[0,0,-1,1]

def solution(graph):
    
    # INIT
    N=len(graph)
    M=len(graph[0])
    visited=[[False]*M for _ in range(N)]
    distance=[[-1]*M for _ in range(N)]
    
    
    # BFS
    queue=deque()
    queue.append((0,0))
    visited[0][0]=True
    distance[0][0]=1
    while queue:
        cx,cy=queue.popleft()
        # base-condition
        if cx==N-1 and cy==M-1:
            return distance[cx][cy]
        for i in range(4):
            nx=cx+dx[i]
            ny=cy+dy[i]
            if 0<=nx<N and 0<=ny<M:
                if not visited[nx][ny] and graph[nx][ny]==1:
                    visited[nx][ny]=True
                    distance[nx][ny]=distance[cx][cy]+1
                    queue.append((nx,ny))
    return -1

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;
class Solution {
    public int solution(int[][] maps) {
        final int N=maps.length;
        final int M=maps[0].length;
        
        boolean[][] visited=new boolean[N][M];
        int[][] dist=new int[N][M];
        
        int[] dx={-1,1,0,0};
        int[] dy={0,0,-1,1};
        
        Deque<int[]> queue=new ArrayDeque<>();
        queue.addLast(new int[]{0,0});
        visited[0][0]=true;
        dist[0][0]=1;
        
        while(!queue.isEmpty()){
            int []cur=queue.pollFirst();
            int cx=cur[0];
            int cy=cur[1];
            if (cx==N-1&&cy==M-1){
                return dist[cx][cy];
            }
            for (int i=0;i<4;i++){
                int nx=cx+dx[i];
                int ny=cy+dy[i];
                if (nx>=0&&nx<N&&ny>=0&&ny<M) {
                    if (maps[nx][ny]==1&&!visited[nx][ny]){
                        visited[nx][ny]=true;
                        dist[nx][ny]=dist[cx][cy]+1;
                        queue.addLast(new int[] {nx,ny});
                    }
                }
            }   
        }
        return -1;
    }
}